### DISCUSSION TITLE
Can't run PCA on merged data

### DISCUSSION THREAD

**Discussion by GreyRockIQ** [satijalab/seurat#5498] [Q&A] [answered]
Hello
I couldn't able to RunPCA on merged data and shows following errors. I have done FindVariableFeatures but still going on like this..

Error in PrepDR(object = object, features = features, verbose = verbose) : 
Variable features haven't been set. Run FindVariableFeatures() or provide a vector of feature names.

Any suggestions?
Thanks
GreyRock

**Comment by mannakade** [comment]
not a dev but hopefully I can help a bit!
It's looking for the vector that holds the features data.
When you have the FindVariableFeatures() line what do you have inside the parenthesis?

**Reply by GreyRockIQ** [reply]
Hi @mannakade 
here is the FindVariableFeatures() code have inside the parenthesis

Warning: The following arguments are not used: nfeatures
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|

Thanks
GreyRock

**Comment by samuel-marsh** [comment, marked as answer]
Can you please post full code that you are running?

**Reply by GreyRockIQ** [reply]
Hi @samuel-marsh 
Here is the code that was run

test.combined <- RunPCA(test.combined, features = VariableFeatures(object = test.combined))

Thanks
GreyRock

**Reply by samuel-marsh** [reply]
Hi,

but where is the code where you ran FindVariableFeatures first?

**Reply by GreyRockIQ** [reply]
Hi @samuel-marsh 
Here is the code
test.combined <- FindVariableFeatures(test.combined, selection.method = "vst", nfeatures = 2000)

Thanks
GreyRock

**Reply by samuel-marsh** [reply]
Try restarting and running things through again.  You shouldn’t need to specify features in RunPCA command.  By default it will use the variable features.

if that doesn’t work please post the full analysis code from the start.

thanks,
Sam

**Reply by GreyRockIQ** [reply]
Thanks @samuel-marsh 
It's worked.
Thanks
GreyRock

**Comment by qimiaonnegliguo** [comment]
> Thanks @samuel-marsh It's worked. Thanks GreyRock
Sorry to bother you. I also use 'merge' function to merge several sample and I wonder if I need to use 'NormalizeData' function before find variable features and run PCA? Thank you very much.

**Comment by RajneeshSrivastava** [comment]
Hi @samuel-marsh  and @GreyRockIQ 

I merged 56 SCTransformed samples (multi-tissue) along with sct normalized data (merge.data = TRUE) and ran PCA on this merged.object. Please confirm if my approach is correct.

Here is my code:
```
order=c("S1","S2","S3",............,"S56")  # for ids
sample.list=list(S2,S3,........,S56)
merged.object=merge(S1, y=sample.list, add.cell.ids = order, merge.data = TRUE)

merged.object
An object of class Seurat 
89591 features across 260259 samples within 2 assays 
Active assay: SCT (39819 features, 0 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
```
### Tried calling "FindVariableFeatures" with SCT assay 
```
merged.object <- FindVariableFeatures(merged.object, assay = "SCT", selection.method = "vst", nfeatures = 3000)
Error: SCT assay is comprised of multiple SCT models. To change the variable features, please set manually with VariableFeatures

merged.object
An object of class Seurat 
89591 features across 260259 samples within 2 assays 
Active assay: SCT (39819 features, 0 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA

merged.object <- RunPCA(merged.object, assay = "SCT") # Ofcourse will not run
Error in PrepDR(object = object, features = features, verbose = verbose) : 
  Variable features haven't been set. Run FindVariableFeatures() or provide a vector of feature names.
```
### Extracted the variable features as [suggested](https://github.com/satijalab/seurat/issues/6185#issuecomment-1185832747) 
```
VariableFeatures(merged.object[["SCT"]]) <- rownames(merged.object[["SCT"]]@scale.data)

merged.object
An object of class Seurat 
89591 features across 260259 samples within 2 assays 
Active assay: SCT (39819 features, 6785 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA

merged.object <- RunPCA(object = merged.object, verbose = FALSE)

merged.object
An object of class Seurat 
89591 features across 260259 samples within 2 assays 
Active assay: SCT (39819 features, 6785 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reduction calculated: pca

```
Is this ok to run PCA with 6785 variable features?
